File: crates/termoso-mobile/src/presence.rs
//! Team presence on mobile: which team-vault hosts this phone is connected
//! to, and who else is on the team's hosts.
//!
//! Every terminal, SFTP browser and tunnel registers a [`Slot`] when it is
//! opened for a saved host; the slot reports "connected" / "gone" and the
//! [`Tracker`] turns the live set into the snapshot the sync engine sends.
//! Only routing metadata leaves the phone (vault, host, protocol, start
//! time); local- and personal-vault hosts are filtered out here.

use std::collections::hash_map::Entry;
use std::collections::{HashMap, HashSet};
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, Mutex, Weak};

use chrono::{DateTime, Utc};
use termoso_core::model::Host;
use termoso_core::store::{LocalVaultKind, Store};
use termoso_core::sync::SyncEngine;
use termoso_proto::team::{PresenceEntry, PresenceSession, TeamPresence};
use uuid::Uuid;

struct Live {
    host_id: Uuid,
    protocol: &'static str,
    since: DateTime<Utc>,
}

/// This device's open connections, shared by every session object and the
/// account runtime.
pub(crate) struct Tracker {
    store: Arc<Store>,
    live: Mutex<HashMap<u64, Live>>,
    next: AtomicU64,
    engine: Mutex<Weak<SyncEngine>>,
}

impl Tracker {
    pub(crate) fn new(store: Arc<Store>) -> Arc<Self> {
        Arc::new(Self {
            store,
            live: Mutex::new(HashMap::new()),
            next: AtomicU64::new(1),
            engine: Mutex::new(Weak::new()),
        })
    }

    /// Publish through `engine` from now on (and push the current set once).
    pub(crate) fn attach(&self, engine: &Arc<SyncEngine>) {
        *self.engine.lock().unwrap_or_else(|p| p.into_inner()) = Arc::downgrade(engine);
        self.publish();
    }

    /// A connection to `host_id` that will report when it is up and down.
    pub(crate) fn slot(self: &Arc<Self>, host_id: Uuid, protocol: &'static str) -> Slot {
        Slot {
            tracker: self.clone(),
            key: self.next.fetch_add(1, Ordering::Relaxed),
            host_id,
            protocol,
        }
    }

    fn connected(&self, key: u64, host_id: Uuid, protocol: &'static str) {
        let changed = {
            let mut live = self.live.lock().unwrap_or_else(|p| p.into_inner());
            match live.entry(key) {
                Entry::Occupied(_) => false,
                Entry::Vacant(slot) => {
                    slot.insert(Live {
                        host_id,
                        protocol,
                        since: Utc::now(),
                    });
                    true
                }
            }
        };
        if changed {
            self.publish();
        }
    }

    fn gone(&self, key: u64) {
        let removed = self
            .live
            .lock()
            .unwrap_or_else(|p| p.into_inner())
            .remove(&key)
            .is_some();
        if removed {
            self.publish();
        }
    }

    /// Team-vault sessions open on this device right now.
    pub(crate) fn collect(&self) -> Vec<PresenceSession> {
        let team_vaults: HashSet<Uuid> = self
            .store
            .vaults()
            .unwrap_or_default()
            .into_iter()
            .filter(|v| v.kind == LocalVaultKind::Team)
            .map(|v| v.id)
            .collect();
        if team_vaults.is_empty() {
            return Vec::new();
        }
        let live = self.live.lock().unwrap_or_else(|p| p.into_inner());
        live.values()
            .filter_map(|l| {
                let host = self.store.require::<Host>(l.host_id).ok()?;
                team_vaults
                    .contains(&host.vault_id)
                    .then(|| PresenceSession {
                        vault_id: host.vault_id,
                        host_id: l.host_id,
                        protocol: l.protocol.to_string(),
                        since: l.since,
                    })
            })
            .collect()
    }

    fn publish(&self) {
        let engine = self
            .engine
            .lock()
            .unwrap_or_else(|p| p.into_inner())
            .upgrade();
        if let Some(engine) = engine {
            engine.set_presence(self.collect());
        }
    }
}

/// One connection's registration. Dropping it (the session object going
/// away) counts as "gone".
pub(crate) struct Slot {
    tracker: Arc<Tracker>,
    key: u64,
    host_id: Uuid,
    protocol: &'static str,
}

impl Slot {
    /// The connection is fully up (shell open / SFTP ready / tunnel bound).
    pub(crate) fn connected(&self) {
        self.tracker
            .connected(self.key, self.host_id, self.protocol);
    }

    /// The connection ended, failed or is reconnecting.
    pub(crate) fn gone(&self) {
        self.tracker.gone(self.key);
    }
}

impl Drop for Slot {
    fn drop(&mut self) {
        self.gone();
    }
}

// ---- what teammates see -------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, uniffi::Record)]
pub struct PresenceSessionCard {
    pub vault_id: String,
    pub host_id: String,
    /// `ssh`, `mosh`, `telnet`, `sftp`, `forward`.
    pub protocol: String,
    /// RFC 3339.
    pub since: String,
}

/// One teammate device and what it is connected to.
#[derive(Debug, Clone, PartialEq, Eq, uniffi::Record)]
pub struct PresenceEntryCard {
    pub user_id: String,
    pub email: String,
    pub display_name: Option<String>,
    pub device_id: String,
    pub device_name: String,
    /// `android`, `linux`, `windows`, `macos`, …
    pub platform: String,
    pub sessions: Vec<PresenceSessionCard>,
    /// RFC 3339.
    pub seen_at: String,
    /// This is the signed-in user (possibly another of their devices).
    pub me: bool,
}

/// Who is on the team's hosts right now. `entries` is empty while the team
/// has presence switched off.
#[derive(Debug, Clone, PartialEq, Eq, uniffi::Record)]
pub struct TeamPresenceCard {
    pub enabled: bool,
    pub entries: Vec<PresenceEntryCard>,
}

pub(crate) fn card(p: TeamPresence, me: Option<Uuid>) -> TeamPresenceCard {
    TeamPresenceCard {
        enabled: p.enabled,
        entries: p.entries.into_iter().map(|e| entry_card(e, me)).collect(),
    }
}

fn entry_card(e: PresenceEntry, me: Option<Uuid>) -> PresenceEntryCard {
    PresenceEntryCard {
        me: me == Some(e.user_id),
        user_id: e.user_id.to_string(),
        email: e.email,
        display_name: e.display_name,
        device_id: e.device_id.to_string(),
        device_name: e.device_name,
        platform: e.platform,
        sessions: e
            .sessions
            .into_iter()
            .map(|s| PresenceSessionCard {
                vault_id: s.vault_id.to_string(),
                host_id: s.host_id.to_string(),
                protocol: s.protocol,
                since: s.since.to_rfc3339(),
            })
            .collect(),
        seen_at: e.seen_at.to_rfc3339(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use termoso_crypto::keys::SymmetricKey;
    use termoso_proto::vault::VaultRole;

    fn host(store: &Store, vault: Uuid, label: &str) -> Uuid {
        store
            .insert(
                vault,
                &Host {
                    label: label.into(),
                    address: label.into(),
                    ..Default::default()
                },
            )
            .expect("host")
    }

    #[test]
    fn team_hosts_only_and_slots_track_lifecycle() {
        let store = Arc::new(Store::open_in_memory(SymmetricKey::generate()).expect("store"));
        let local = store.local_vault().expect("local vault").id;
        let team = Uuid::new_v4();
        store
            .upsert_vault(
                team,
                LocalVaultKind::Team,
                "Ops",
                Some(Uuid::new_v4()),
                VaultRole::Editor,
                Some(&SymmetricKey::generate()),
                1,
            )
            .expect("team vault");
        let local_host = host(&store, local, "a.local");
        let team_host = host(&store, team, "b.team");

        let tracker = Tracker::new(store);
        let a = tracker.slot(local_host, "ssh");
        let b = tracker.slot(team_host, "ssh");
        let b2 = tracker.slot(team_host, "sftp");
        a.connected();
        b.connected();
        assert_eq!(
            tracker.collect().len(),
            1,
            "local-vault host is never reported"
        );
        b2.connected();
        let mut protocols: Vec<String> =
            tracker.collect().into_iter().map(|s| s.protocol).collect();
        protocols.sort();
        assert_eq!(protocols, ["sftp", "ssh"]);
        b.gone();
        assert_eq!(tracker.collect().len(), 1);
        drop(b2);
        assert!(
            tracker.collect().is_empty(),
            "dropping a slot counts as gone"
        );
        b.connected();
        b.connected();
        assert_eq!(tracker.collect().len(), 1, "connected is idempotent");
    }
}

File: apps/desktop/src/hosts/presence.ts
import type { PresenceEntry, PresenceSession, TeamPresence, Uuid } from "@/ipc/types";
import { tr, msg } from "@/i18n";

/** One teammate's device on one host: the sessions it has open there. */
export interface HostViewer {
  userId: Uuid;
  email: string;
  displayName: string | null;
  avatar: string | null;
  deviceId: Uuid;
  deviceName: string;
  platform: string;
  /** Distinct protocols, in the order they were opened. */
  protocols: string[];
  /** Earliest `since` across the device's sessions on this host. */
  since: string;
  /** This is the signed-in account (any of its devices). */
  me: boolean;
}

/** Who is on which host, keyed by host id. Devices are listed oldest connection first. */
export function viewersByHost(
  presence: TeamPresence | undefined,
  myUserId: Uuid | null,
): Map<Uuid, HostViewer[]> {
  const out = new Map<Uuid, HostViewer[]>();
  if (!presence?.enabled) return out;
  for (const e of presence.entries) {
    for (const [hostId, sessions] of byHost(e.sessions)) {
      const list = out.get(hostId) ?? [];
      list.push(viewer(e, sessions, myUserId));
      out.set(hostId, list);
    }
  }
  for (const list of out.values()) list.sort((a, b) => a.since.localeCompare(b.since));
  return out;
}

function byHost(sessions: PresenceSession[]): Map<Uuid, PresenceSession[]> {
  const m = new Map<Uuid, PresenceSession[]>();
  for (const s of sessions) {
    const list = m.get(s.host_id) ?? [];
    list.push(s);
    m.set(s.host_id, list);
  }
  return m;
}

function viewer(e: PresenceEntry, sessions: PresenceSession[], myUserId: Uuid | null): HostViewer {
  const sorted = [...sessions].sort((a, b) => a.since.localeCompare(b.since));
  const protocols: string[] = [];
  for (const s of sorted) if (!protocols.includes(s.protocol)) protocols.push(s.protocol);
  return {
    userId: e.user_id,
    email: e.email,
    displayName: e.display_name ?? null,
    avatar: e.avatar ?? null,
    deviceId: e.device_id,
    deviceName: e.device_name,
    platform: e.platform,
    protocols,
    since: sorted[0]?.since ?? e.seen_at,
    me: myUserId !== null && e.user_id === myUserId,
  };
}

/** Display name when set, else the email. */
export const viewerName = (v: HostViewer) =>
  v.displayName !== null && v.displayName.trim() !== "" ? v.displayName : v.email;

/** Distinct people (not devices) among the viewers, in first-seen order. */
export function distinctPeople(viewers: HostViewer[]): HostViewer[] {
  const seen = new Set<Uuid>();
  return viewers.filter((v) => (seen.has(v.userId) ? false : (seen.add(v.userId), true)));
}

/** `2 min`, `1 h 05 min`, `3 d 2 h` — how long a connection has been open. */
export function connectedFor(sinceIso: string, now = Date.now()): string {
  const t = Date.parse(sinceIso);
  if (Number.isNaN(t)) return "";
  const s = Math.max(0, Math.floor((now - t) / 1000));
  if (s < 60) return "just now";
  const m = Math.floor(s / 60);
  if (m < 60) return `${m} min`;
  const h = Math.floor(m / 60);
  if (h < 24) return `${h} h ${String(m % 60).padStart(2, "0")} min`;
  const d = Math.floor(h / 24);
  return `${d} d ${h % 24} h`;
}

const PROTOCOL_LABEL: Record<string, string> = {
  ssh: "SSH",
  mosh: "Mosh",
  telnet: "Telnet",
  sftp: "SFTP",
  forward: msg("Port forwarding"),
  serial: "Serial",
};

export const protocolLabel = (p: string) => tr(PROTOCOL_LABEL[p] ?? p.toUpperCase());

const PLATFORM_LABEL: Record<string, string> = {
  windows: "Windows",
  linux: "Linux",
  macos: "macOS",
  android: "Android",
  ios: "iOS",
  web: "Web",
  cli: "CLI",
};

export const platformLabel = (p: string) => PLATFORM_LABEL[p] ?? p;

/** `Alice`, `Alice and Bob`, `Alice, Bob and 2 others` — for a card tooltip. */
export function viewersSummary(viewers: HostViewer[]): string {
  const names = distinctPeople(viewers).map((v) => (v.me ? "You" : viewerName(v)));
  if (names.length === 0) return "";
  if (names.length === 1) return names[0] ?? "";
  if (names.length === 2) return `${names[0]} and ${names[1]}`;
  const rest = names.length - 2;
  return `${names[0]}, ${names[1]} and ${rest} other${rest === 1 ? "" : "s"}`;
}
